lib/text_diff.ml

```diff
@@ -119,24 +126,229 @@ let process_atomic_diff (diff : (_, _ Diff.atomic_modification) Diff.t) name
+and process_modified_type_diff name type_kind type_privacy type_manifest =
+  let type_header_diff =
+    process_type_header_diff name type_privacy type_manifest type_kind
+  in
+  let type_kind_diff = process_type_kind_diff type_kind in
+  order_type_diffs type_header_diff type_kind_diff
+
+and order_type_diffs type_header_diff type_kind_diff =
+  match (type_header_diff, type_kind_diff) with
+  | `Same _, `Same _ -> assert false
+  | `Same type_header, `Atomic_change type_kind_change
+  | `Same type_header, `Compound_change type_kind_change ->
+      type_header @ type_kind_change
+  | `Atomic_change type_header_change, `Same type_kind ->
+      type_header_change @ type_kind
+  | `Atomic_change type_header_change, `Atomic_change type_kind_change ->
+      List.hd type_header_change :: List.hd type_kind_change
+      :: List.tl type_header_change
+      @ List.tl type_kind_change
+  | `Atomic_change type_header_change, `Compound_change type_kind_change ->
+      type_header_change @ type_kind_change
+
+and process_type_header_diff name type_privacy type_manifest type_kind =
+  match (type_privacy, type_manifest, type_kind) with
+  | Either.Left private_flag, Either.Left te_opt, Either.Right (Record_tk _)
+  | Either.Left private_flag, Either.Left te_opt, Either.Right (Variant_tk _)
+  | ( Either.Left private_flag,
+      Either.Left te_opt,
+      Either.Right
+        (Atomic_tk
+          {
+            reference = Types.Type_record (_, _);
+            current = Types.Type_record (_, _);
+          }) )
+  | ( Either.Left private_flag,
+      Either.Left te_opt,
+      Either.Right
+        (Atomic_tk
+          {
+            reference = Types.Type_record (_, _);
+            current = Types.Type_variant (_, _);
+          }) )
+  | ( Either.Left private_flag,
+      Either.Left te_opt,
+      Either.Right
+        (Atomic_tk
+          { reference = Types.Type_record (_, _); current = Types.Type_open }) )
+  | ( Either.Left private_flag,
+      Either.Left te_opt,
+      Either.Right
+        (Atomic_tk
+          {
+            reference = Types.Type_variant (_, _);
+            current = Types.Type_record (_, _);
+          }) )
+  | ( Either.Left private_flag,
+      Either.Left te_opt,
+      Either.Right
+        (Atomic_tk
+          {
+            reference = Types.Type_variant (_, _);
+            current = Types.Type_variant (_, _);
+          }) )
+  | ( Either.Left private_flag,
+      Either.Left te_opt,
+      Either.Right
+        (Atomic_tk
+          { reference = Types.Type_variant (_, _); current = Types.Type_open })
+    )
+  | ( Either.Left private_flag,
+      Either.Left te_opt,
+      Either.Right
+        (Atomic_tk
+          { reference = Types.Type_open; current = Types.Type_record (_, _) }) )
+  | ( Either.Left private_flag,
+      Either.Left te_opt,
+      Either.Right
+        (Atomic_tk
+          { reference = Types.Type_open; current = Types.Type_variant (_, _) })
```

Comment:
Also, I'm pretty sure this pattern-match can be made more explicit: you are only matching on a specific part, all the rest is duplicated. You can do something like this:

```ocaml
 | ( Either.Left private_flag,
      Either.Left te_opt,
      Either.Right
        (Atomic_tk
          (CASE1 | CASE2 | CASE3) ->
```